### PR TITLE
LASCOMap: clear rotation metadata for Helioviewer JPEG2000 images

### PR DESCRIPTION
**Pull request by johan12345**
As discussed in #4553, SOHO/LASCO coronagraph images in JPEG2000 format by Helioviewer.org are already rotated so that solar north is pointing up, but they still have the rotation metadata (CROTA2 keyword) from the original FITS file, which no longer applies here.

As suggested by @ayshih, I have added some code to clear the rotation metadata for files coming from Helioviewer.

### CODE REVIEW

File: sunpy/map/sources/tests/test_lasco_source.py [hunk 1]
```diff
@@ -14,6 +15,10 @@
 fitspath = glob.glob(os.path.join(path, "lasco_c2_25299383_s.fts"))
 lasco = Map(fitspath)
 
+jp2path = glob.glob(os.path.join(path, "2013_05_13__16_54_06_137__SOHO_LASCO_C3_white-light.jp2"))
+lasco_helioviewer = Map(jp2path)
```
Comment (nabobalis) [inline sunpy/map/sources/tests/test_lasco_source.py:19]: I know you are following the style of this file but ideally this should be a pytest fixture like so:
```python
@pytest.fixture()
def lasco_helioviewer():
    return Map(jp2path)
```

Then you can pass "lasco_helioviewer" into the test you wrote.

Comment (johan12345) [reply sunpy/map/sources/tests/test_lasco_source.py:19]: sure, that makes sense. Should I do this only for `lasco_helioviewer` or also adjust the `lasco` file to use a fixture?

Comment (nabobalis) [reply sunpy/map/sources/tests/test_lasco_source.py:19]: If you can that would be great.

Comment (johan12345) [reply sunpy/map/sources/tests/test_lasco_source.py]: done in 8a458bc.

Comment (nabobalis) [reply sunpy/map/sources/tests/test_lasco_source.py]: Thanks.

**Comment by nabobalis** [comment]
Looks good to me but I will let someone with more knowledge of maps to approve this.

**Review by dstansby — CHANGES_REQUESTED** [review]
Thanks for the PR! I think this is broadly the right approach, but we need to protect against users subsequently rotating the map, saving it, and loading it in to map again. In this case, the rotation metadata should be preserved.

In other words we need to find a robust way of knowing whether a file is a completely untouched file that has been loaded from a helioviewer file, and I don't think just checking for 'helioviewer' in the metadta is sufficient.

**Comment by johan12345** [comment]
I just tested what happens in this case:

```python
# download a Helioviewer image
date = dt.datetime(2013, 5, 13, 13, 00)
hv = helioviewer.HelioviewerClient()
file = hv.download_jp2(date, observatory='SOHO', instrument='EIT', measurement='171')
data, header = read_file(file)[0]
map = Map(data, header)
print(map.rotation_matrix)
# > [[ 1. -0.] [ 0.  1.]]

map_rot = map.rotate(90 * degree)
print(map_rot.rotation_matrix)
# > [[ 6.123234e-17  1.000000e+00] [-1.000000e+00  6.123234e-17]]

map_rot.save('test.fts')
# WARNING: SunpyUserWarning: The meta key history is not valid ascii, dropping from the FITS header [sunpy.io.fits]
# WARNING: SunpyUserWarning: The meta key helioviewer is too long, dropping from the FITS header (maximum allowed key length is 8 characters). [sunpy.io.fits]

map_rot_reload = Map('test.fts')
print(map_rot_reload.rotation_matrix)
# > [[ 6.123234e-17  1.000000e+00] [-1.000000e+00  6.123234e-17]]
```

So, at the moment, the `helioviewer` key is dropped from the metadata when saving to FITS as it is too long. So, it works correctly - even though that's obviously not a perfect solution.

But I think even if the `helioviewer` key would not be dropped it would still work, because the `rotate()` function removes the `CROTAn` headers and adds `pc1_1, pc1_2, pc2_1, pc_2_2` headers instead to save the rotation matrix.

**Comment by ayshih** [comment]
> So, at the moment, the `helioviewer` key is dropped from the metadata when saving to FITS as it is too long.

I think there's an argument for explicitly dropping the `helioviewer` keyword when we start mucking around with the metadata.

> In other words we need to find a robust way of knowing whether a file is a completely untouched file that has been loaded from a helioviewer file, and I don't think just checking for 'helioviewer' in the metadta is sufficient.

> But I think even if the `helioviewer` key would not be dropped it would still work, because the `rotate()` function removes the `CROTAn` headers and adds `pc1_1, pc1_2, pc2_1, pc_2_2` headers instead to save the rotation matrix.

The detection in this PR can be made more robust by checking for the existence of all three `CROTAn` keywords and performing the cleanup only if they are *all* there.  If after this cleanup code is run, some user wants to avoid using `.rotate()` and instead chooses to directly modify `CROTA2`, that would not trigger the cleanup code again as long as the user doesn't *also* inject both `CROTA` and `CROTA1` (which aren't supported keywords anyway).

**Comment by johan12345** [comment]
@ayshih 
> The detection in this PR can be made more robust by checking for the existence of all three CROTAn keywords and performing the cleanup only if they are all there. If after this cleanup code is run, some user wants to avoid using .rotate() and instead chooses to directly modify CROTA2, that would not trigger the cleanup code again as long as the user doesn't also inject both CROTA and CROTA1 (which aren't supported keywords anyway).

Okay, that makes sense. I have added a check for all three CROTAn keywords exist before removing them.

**Comment by Cadair** [comment]
Have we thrown rocks at helioviewer about this? it's naughty they are shipping data with lies in the header :p

**Comment by ayshih** [comment]
@wafels gave a detailed comment in the associated issue (https://github.com/sunpy/sunpy/issues/4553#issuecomment-706196247).  In short, they preserved the original FITS keyword values instead of updating the metadata be consistent with modifications such as de-rotation because the files were not designed to be reused as science data.

**Comment by wtbarnes** [comment]
If these files are not meant for science and are really just for a quick look , should we raise a warning indicating as much?

**Comment by johan12345** [comment]
@wtbarnes 
> If these files are not meant for science and are really just for a quick look , should we raise a warning indicating as much?

Well, that might make sense, however if I understand correctly they are the only LASCO images available within SunPy that are already processed (background subtraction etc.) - the raw images that can be downloaded from VSO with `Fido` are not of much use for science either...

**Comment by ayshih** [comment]
I think raising a warning would be silly.  The Helioviewer files aren't fundamentally flawed and are still worthwhile for science, even with the caveats raised by @wafels.  The files simply intentionally preserve metadata that isn't consistent with the data because their design hadn't anticipated that someone (e.g., SunPy) would blindly rely on that metadata as opposed to referring to it as a historical record.

**Review by dstansby — COMMENTED** [review]
Looks good to me now; could you rebase this on to the current master branch, and hopefully that will trigger a run of the tests?

**Comment by johan12345** [comment]
Sure, I have rebased, now waiting for the CI to run the tests :)

**Comment by johan12345** [comment]
Some tests still seem to be failing, also the one that I added:
```
test_helioviewer_rotation

failed on setup with "OSError: Failed to read d:\a\1\s\.tox\py37\lib\site-packages\sunpy\data\test\2013_05_13__16_54_06_137__SOHO_LASCO_C3_white-light.jp2."
```
I'm not sure why reading this file fails, it works for me locally. It also doesn't seem to be failing in all of the test environments.

**Comment by ayshih** [comment]
The reading is failing because the Windows environment doesn't have OpenJPEG for Glymur.
```
E           RuntimeError: You must have a version of OpenJPEG at least as high as 2.3.0 before you can read JPEG2000 images with glymur.  Your version is 0.0.0
```
I don't remember whether there's a workaround.

(The other test failures can be ignored.)

**Comment by ayshih** [comment]
Oh, right, you can decorate the test with `@skip_glymur` so that the test will be skipped if glymur isn't there or doesn't see OpenJPEG

**Comment by johan12345** [comment]
okay, let's see whether it works now

**Comment by johan12345** [comment]
Now all tests are passing :)

**Review by dstansby — APPROVED** [review]

File: sunpy/map/sources/soho.py [hunk 2]
```diff
@@ -123,6 +124,17 @@ def __init__(self, data, header, **kwargs):
         self.plot_settings['norm'] = ImageNormalize(
             stretch=source_stretch(self.meta, PowerStretch(0.5)), clip=False)
 
+        # For Helioviewer images, clear rotation metadata, as these have already been rotated.
+        # Also check that all CROTAn keywords exist to make sure that it's an untouched
+        # Helioviewer file.
+        if 'helioviewer' in self.meta \
+                and 'crota' in self.meta and 'crota1' in self.meta and 'crota2' in self.meta:
```
Comment (Cadair) [inline sunpy/map/sources/soho.py:131]: I have a hatred of line continuation characters :stuck_out_tongue: 
```suggestion
        if ('helioviewer' in self.meta 
                and 'crota' in self.meta and 'crota1' in self.meta and 'crota2' in self.meta):
```

**Review by Cadair — APPROVED** [review]

File: sunpy/map/sources/soho.py (hunk 2 above)
Comment (johan12345) [reply sunpy/map/sources/soho.py]: Haha, no problem 😅 Replaced in f6496e2.

**Comment by Cadair** [comment]
@johan12345 just so you know I have rebased this to try and fix the CI.

**Comment by Cadair** [comment]
Thanks a lot @johan12345 :rocket:

**Comment by johan12345** [comment]
Great, thank you for your help 🌞

**Pull request merged**